### PR TITLE
[FIX] mail: leave sub-thread crashes for non-admin user

### PR DESCRIPTION
**Pull request by maki-odoo**
This commit fixes the following issue in the sub-thread behavior.
Steps to reproduce:
- Create a thread as non admin user and add another member
- Create a sub-thread
- Go to the parent thread and type `/leave` command
- You see the access error

### CODE REVIEW

File: addons/mail/models/discuss/discuss_channel_member.py [hunk 1]
```diff
@@ -217,9 +217,13 @@ def unlink(self):
...
-                lambda m: m.partner_id in set(parent.channel_member_ids.partner_id)
-                if m.partner_id
-                else m.guest_id in set(parent.channel_member_ids.guest_id)
+                lambda m: (
+                    m.partner_id in self.partner_id
+                    and m.partner_id in set(parent.channel_member_ids.partner_id)
+                    if m.partner_id
+                    else m.guest_id in self.guest_id
+                         and m.guest_id in set(parent.channel_member_ids.guest_id)
+                )
```
Comment (alexkuhn) [inline addons/mail/models/discuss/discuss_channel_member.py:226]: Ok, code was not taking `self` into account so it was `_action_unfollow`ing all members. This limits the `_action_unfollow` to `self` members, as it should.

File: addons/mail/models/discuss/discuss_channel.py [hunk 2]
```diff
@@ -1134,7 +1134,10 @@ def _create_sub_channel(self, from_message_id=None, name=None):
             message = self.env["mail.message"].search([("id", "=", from_message_id)])
         sub_channel = self.create(
             {
-                "channel_member_ids": [Command.create({"partner_id": self.env.user.partner_id.id})],
+                "channel_member_ids": [
+                    Command.create({"partner_id": partner.id})
+                    for partner in self.channel_member_ids.partner_id | self.env.user.partner_id
+                ],
```
Comment (alexkuhn) [inline addons/mail/models/discuss/discuss_channel.py:1140]: This could be inefficient, e.g. if we have a channel with ten thousands of members, creating a sub thread is costly.
Any user with any message can make a sub-thread so it'd be nice for sub-thread creation to stay cheap.
When accessing the thread, we could create channel member on the fly if the user is a member of `parent_channel_id`.

I'm not sure whether the presence of channel member is important for other things, e.g. being notified with `@mention`. But if there's no such problem, would be nice to avoid creating a member record if it's not mandatory (again to save space when we could).

**Review by alexkuhn — COMMENTED** [review]
LGTM otherwise

**Review by alexkuhn — APPROVED** [review]
@robodoo r+

File: addons/mail/models/discuss/discuss_channel_member.py [hunk 3]
```diff
@@ -214,13 +214,14 @@ def unlink(self):
...
-                lambda m: m.partner_id in set(parent.channel_member_ids.partner_id)
-                if m.partner_id
-                else m.guest_id in set(parent.channel_member_ids.guest_id)
-            )
+        sub_members = self.env["discuss.channel.member"].search(
+            [
+                "|",
+                ("partner_id", "in", self.partner_id.ids),
+                ("guest_id", "in", self.guest_id.ids),
+                ("channel_id", "in", self.channel_id.sub_channel_ids.ids),
```
Comment (seb-odoo) [inline addons/mail/models/discuss/discuss_channel_member.py:222]: This is not correct. Each channel should be mapped to specific guests or partners.

If unlink is called on partner 1 of channel 2 and partner 2 of channel 1, this will incorrectly also return partner 1 of channel 2 and partner 2 of channel 2.

**Comment by alexkuhn** [comment]
@robodoo r-

File: addons/mail/models/discuss/discuss_channel_member.py (hunk 3 above)
Comment (maki-odoo) [reply addons/mail/models/discuss/discuss_channel_member.py]: Done

**Review by alexkuhn — APPROVED** [review]
LGTM, @seb-odoo for double-check

File: addons/mail/tests/discuss/test_discuss_sub_channels.py [hunk 4]
```diff
@@ -31,26 +31,27 @@ def test_01_gc_unpin_outdated_sub_channels(self):
...
 
-    def test_02_sub_channel_members_sync_with_parent(self):
-        parent = self.env["discuss.channel"].create({"name": "General"})
-        parent.action_unfollow()
-        self.assertFalse(any(m.is_self for m in parent.channel_member_ids))
+    def test_02_sub_channel_members(self):
+        bob_user = new_test_user(self.env, "bob_user", groups="base.group_user")
+        baz_user = new_test_user(self.env, "baz_user", groups="base.group_user")
+        parent = self.env["discuss.channel"].with_user(bob_user).create({"name": "General"})
+        self.assertTrue(any(m.is_self for m in parent.channel_member_ids))
```
Comment (seb-odoo) [inline addons/mail/tests/discuss/test_discuss_sub_channels.py:38]: While we have diff on them, all of these should be `assertIn` or `assertNotIn`.

File: addons/mail/models/discuss/discuss_channel_member.py [hunk 5]
```diff
@@ -212,16 +212,21 @@ def write(self, vals):
...
+                lambda m: (m.partner_id.id, m.guest_id.id)
+        ).items():
+            domain = expression.OR(
+                [
+                    domain,
+                    [
+                        ("partner_id" if p_id else "guest_id", "=", p_id or g_id),
+                        ("channel_id", "in", rec.channel_id.sub_channel_ids.ids),
+                    ],
+                ]
```
Comment (seb-odoo) [inline addons/mail/models/discuss/discuss_channel_member.py:227]: Now that I see it, I'm wondering if it's worth it to group them, it's probably easier to make subdomains for each member directly.

```suggestion
        domains = [
            [
                ("partner_id", "=", member.partner_id.id),
                ("guest_id", "=", member.guest_id.id),
                ("channel_id", "in", member.channel_id.sub_channel_ids.ids),
            ]
            for member in self
        ]
```

File: addons/mail/models/discuss/discuss_channel_member.py [hunk 6]
```diff
@@ -212,16 +212,21 @@ def write(self, vals):
...
             )
-        for member in sub_members:
+        for member in self.env["discuss.channel.member"].search(domain):
```
Comment (seb-odoo) [inline addons/mail/models/discuss/discuss_channel_member.py:229]:
```suggestion
        for member in self.env["discuss.channel.member"].search(expression.OR(domains)):
```

**Review by seb-odoo — COMMENTED** [review]
It looks like it works as expected, but code could be cleaned up a little bit more.

File: addons/mail/tests/discuss/test_discuss_sub_channels.py [hunk 7]
```diff
@@ -93,3 +94,30 @@ def test_06_initial_message_must_belong_to_parent_channel(self):
...
+            parent_1_baz_member
+            in parent_1.channel_member_ids | parent_1.sub_channel_ids.channel_member_ids
+        )
+        self.assertFalse(
+            parent_2_bob_member
+            in parent_2.channel_member_ids | parent_2.sub_channel_ids.channel_member_ids
+        )
+        self.assertTrue(bob_user.partner_id in parent_1_sub_channel_1.channel_member_ids.partner_id)
+        self.assertTrue(bob_user.partner_id in parent_1_sub_channel_2.channel_member_ids.partner_id)
+        self.assertTrue(baz_user.partner_id in parent_2_sub_channel.channel_member_ids.partner_id)
```
Comment (seb-odoo) [inline addons/mail/tests/discuss/test_discuss_sub_channels.py:123]: The test is missing a guest on a 3rd channel to have full coverage.

File: addons/mail/tests/discuss/test_discuss_sub_channels.py [hunk 8]
```diff
@@ -31,26 +31,27 @@ def test_01_gc_unpin_outdated_sub_channels(self):
...
+        bob_user = new_test_user(self.env, "bob_user", groups="base.group_user")
+        baz_user = new_test_user(self.env, "baz_user", groups="base.group_user")
+        parent = self.env["discuss.channel"].with_user(bob_user).create({"name": "General"})
+        self.assertIn(bob_user.partner_id, parent.channel_member_ids.partner_id)
+        parent.add_members(partner_ids=[baz_user.partner_id.id])
+        sub_channel = parent._create_sub_channel()
+        # Member created for sub channel (_create_sub_channel): should be only
+        # the sub channel creator.
+        self.assertIn(bob_user.partner_id, sub_channel.channel_member_ids.partner_id)
+        self.assertNotIn(baz_user.partner_id, sub_channel.channel_member_ids.partner_id)
```
Comment (seb-odoo) [inline addons/mail/tests/discuss/test_discuss_sub_channels.py:44]: Why is this part changed actually?
The PR is only about fixing `unlink`, why is a test about "create" changed?

Comment (maki-odoo) [reply addons/mail/tests/discuss/test_discuss_sub_channels.py:44]: The description in the test was `Member created for sub channel (_create_sub_channel): should also be created for the parent channel` and it was considering all members in the channel are adding to the sub channel too. The test was passed before because it was using only odoo bot as a member so in adding the sub channel it was only sub channel creator and checking if all members of parent are in the child was always correct.  In the first version of this PR I changed the code to support this behavior but as you said this is not what we want. I changed the test to focus on correct members in the parent that are not automatically added to the child. It could be in a separate PR, but since we had the related discussion here and the test was already changed, I kept it here.

Comment (seb-odoo) [reply addons/mail/tests/discuss/test_discuss_sub_channels.py:44]: I don't think the test was covering that all members in parents are in child.

It was covering that adding member to child also added members to parent, which is necessary to pass ACL check of child (based on parent).

This is what the test should cover and it no longer covers that now.

Comment (seb-odoo) [reply addons/mail/tests/discuss/test_discuss_sub_channels.py:44]: Is it green if you fully revert diff on this test?

Comment (seb-odoo) [reply addons/mail/tests/discuss/test_discuss_sub_channels.py:44]: sync:
- create sub member -> create parent member
- unlink parent member -> unlink sub member
not sync:
- unlink sub member (parent member is not changed, just leaving sub is allowed)
- create parent member (sub member is not changed, just joining parent and not joining sub is allowed)

Comment (seb-odoo) [reply addons/mail/tests/discuss/test_discuss_sub_channels.py:44]: and the coverage of this test (1st step) was:

- have empty channel (at least self not being member)
- create a sub channel (and therefore create a sub channel member)
- by sync mechanism: parent member should be created

Comment (maki-odoo) [reply addons/mail/tests/discuss/test_discuss_sub_channels.py]: I reverted the diff on this test.

**Review by seb-odoo — APPROVED** [review]
@robodoo r+
